languages/java-cpg/src/main/java/de/jplag/java_cpg/transformation/matching/pattern/PatternRegistry.java

```diff
@@ -0,0 +1,118 @@
+package de.jplag.java_cpg.transformation.matching.pattern;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import de.fraunhofer.aisec.cpg.graph.Node;
+import de.jplag.java_cpg.transformation.GraphTransformation;
+import de.jplag.java_cpg.transformation.Role;
+
+/**
+ * The {@link PatternRegistry} saves the {@link NodePattern}s involved in a {@link GraphTransformation} and their
+ * identifiers.
+ */
+public class PatternRegistry {
+    private static final String WILDCARD_PARENT_ID = "wildcardParent#";
+    private final Map<Role, NodePattern<?>> patternByRole;
+    private final Map<NodePattern<?>, Role> roleByPattern;
+
+    /**
+     * A NodePattern that represents the {@link GraphPattern}. If not set, it is the (first) root of the
+     * {@link GraphPattern}.
+     */
+    private NodePattern<Node> representingNode;
+
+    private static final Logger logger = LoggerFactory.getLogger(PatternRegistry.class);
+    private int wildcardCounter;
+
+    /**
+     * Creates a new {@link PatternRegistry}.
+     */
+    public PatternRegistry() {
+        this.patternByRole = new HashMap<>();
+        this.roleByPattern = new HashMap<>();
+    }
+
+    /**
+     * Gets the {@link NodePattern} with the given {@link Role}, cast to the given {@link Node} class.
+     * @param role the role
+     * @return the node pattern associated to the role
+     */
+    public <T extends Node> NodePattern<T> getPattern(Role role, Class<T> targetClass) {
+        NodePattern<?> nodePattern = patternByRole.get(role);
+        if (!targetClass.isAssignableFrom(nodePattern.getRootClass())) {
+            throw new ClassCastException("Pattern %s is incompatible with target class %s".formatted(role, targetClass));
+        }
+        @SuppressWarnings("unchecked")
```

Comment:
Can you address this warning?

Reply:
A problem I had at multiple stages in this project was that I needed to store objects of generic types with different type arguments in a Map, as is the case with `NodePattern<T>` here, and when I retrieve the object, I need the concrete type argument `T` back. So I make sure that the stored object actually has the type argument that I expect, and then cast it to that type.

Now, because I am reasonably sure that nothing can go wrong here, I deliberately suppressed the warning so that nobody would have to wrap their heads about all of this. Should I remove it either way?